### PR TITLE
LibWeb: Implement IDBFactory::cmp

### PR DESCRIPTION
**Pull request by stelar7**
100% passing on
https://wpt.live/IndexedDB/idbfactory_cmp.htm
https://wpt.live/IndexedDB/idbfactory_cmp2.htm
https://wpt.live/IndexedDB/idbfactory_cmp3.htm
https://wpt.live/IndexedDB/idbfactory_cmp4.htm

### CODE REVIEW

File: Libraries/LibWeb/Infra/Strings.cpp [hunk 1]
```diff
@@ -167,4 +167,29 @@ String isomorphic_decode(ReadonlyBytes input)
...
+        return false;
+
+    // 2. If a is a code unit prefix of b, then return true.
+    if (is_code_unit_prefix(a, b))
+        return true;
+
+    // 3. Let n be the smallest index such that the nth code unit of a is different from the nth code unit of b.
+    //    (There has to be such an index, since neither string is a prefix of the other.)
+    u32 n = 0;
+    while (n < a.length() && n < b.length() && a[n] == b[n])
```
Comment (gmta) [inline Libraries/LibWeb/Infra/Strings.cpp:184]: In the spec, `code unit` is typed as an `u16`, but you're comparing `char`s here with `a[n] == b[n]`?

File: Libraries/LibWeb/IndexedDB/IDBFactory.cpp [hunk 2]
```diff
@@ -93,4 +94,24 @@ WebIDL::ExceptionOr<JS::NonnullGCPtr<IDBOpenDBRequest>> IDBFactory::open(String
     return request;
 }
 
+WebIDL::ExceptionOr<i8> IDBFactory::cmp(JS::Value first, JS::Value second)
```
Comment (gmta) [inline Libraries/LibWeb/IndexedDB/IDBFactory.cpp:97]: Please add a spec link here as well

File: Libraries/LibWeb/Infra/Strings.cpp (hunk 1 above)
Comment (stelar7) [reply Libraries/LibWeb/Infra/Strings.cpp]: Good catch, I didnt think about them being u16! 😅

File: Libraries/LibWeb/Infra/Strings.cpp [hunk 3]
```diff
@@ -167,4 +167,36 @@ String isomorphic_decode(ReadonlyBytes input)
...
+    // 1. If b is a code unit prefix of a, then return false.
+    if (is_code_unit_prefix(b, a))
+        return false;
+
+    // 2. If a is a code unit prefix of b, then return true.
+    if (is_code_unit_prefix(a, b))
+        return true;
+
+    auto code_units_a = utf8_to_utf16(a).release_value_but_fixme_should_propagate_errors();
+    auto code_units_b = utf8_to_utf16(b).release_value_but_fixme_should_propagate_errors();
```
Comment (tcl3) [inline Libraries/LibWeb/Infra/Strings.cpp:182]: We tend to use `MUST` for things that can only fail due to OOM nowadays.

**Review by tcl3 — COMMENTED** [review]
> 100% passing on
https://wpt.live/IndexedDB/idbfactory_cmp.htm
https://wpt.live/IndexedDB/idbfactory_cmp2.htm
https://wpt.live/IndexedDB/idbfactory_cmp3.htm
https://wpt.live/IndexedDB/idbfactory_cmp4.htm

This change also leads to an improvement in this test:
* https://wpt.live/IndexedDB/keyorder.htm

Can we import these?

**Comment by stelar7** [comment]
> > 100% passing on
> > [wpt.live/IndexedDB/idbfactory_cmp.htm](https://wpt.live/IndexedDB/idbfactory_cmp.htm)
> > [wpt.live/IndexedDB/idbfactory_cmp2.htm](https://wpt.live/IndexedDB/idbfactory_cmp2.htm)
> > [wpt.live/IndexedDB/idbfactory_cmp3.htm](https://wpt.live/IndexedDB/idbfactory_cmp3.htm)
> > [wpt.live/IndexedDB/idbfactory_cmp4.htm](https://wpt.live/IndexedDB/idbfactory_cmp4.htm)
> 
> This change also leads to an improvement in this test:
> 
> * [wpt.live/IndexedDB/keyorder.htm](https://wpt.live/IndexedDB/keyorder.htm)
> 
> Can we import these?

Imported 👍

**Pull request merged**